src/render.rs

```diff
@@ -15,169 +14,217 @@ use matrix_sdk::{
     },
     PossiblyRedactedExt,
 };
+
 use weechat::Weechat;
 
-/// Rendering function for membership events (joins, leaves, bans, profile
-/// changes, etc).
-// TODO: We should not return raw strings here but something like [Block]
-// where Block is (String, [Tags]). Each Block represents one or several lines
-// which have the same tags.
-pub fn render_membership(
-    member: &SyncStateEvent<MemberEventContent>,
-    sender: &WeechatRoomMember,
-    target: &WeechatRoomMember,
-) -> String {
-    use MembershipChange::*;
-    let change_op = member.membership_change();
-
-    let operation = match change_op {
-        None => "did nothing",
-        Error => "caused an error", // must never happen
-        Joined => "has joined the room",
-        Left => "has left the room",
-        Banned => "was banned by",
-        Unbanned => "was unbanned by",
-        Kicked => "was kicked from the room by",
-        Invited => "was invited to the room by",
-        KickedAndBanned => "was kicked and banned by",
-        InvitationRejected => "rejected the invitation",
-        InvitationRevoked => "had the invitation revoked by",
-        ProfileChanged { .. } => "_",
-        NotImplemented => "performed an unimplemented operation",
-        _ => "caused an unhandled membership change",
-    };
+use crate::room_buffer::WeechatRoomMember;
 
-    fn formatted_name(member: &WeechatRoomMember) -> String {
-        match &member.display_name {
-            Some(display_name) => {
-                format!("{name} {color_delim}({color_reset}{user_id}{color_delim}){color_reset}",
-                        name = display_name,
-                        user_id = &member.user_id,
-                        color_delim = Weechat::color("chat_delimiters"),
-                        color_reset = Weechat::color("reset"))
-            }
+/// The rendered version of an event.
+pub struct RenderedEvent {
+    /// The UNIX timestamp of the event.
+    pub message_timestamp: u64,
+    /// The tags of the event,
+    pub tags: Vec<String>,
+    // TODO we need to allow multiple message/tags pairs here.
+    // This will be useful if a message has a part that we should be able to hide
+    // using a filter on the tags.
+    //
+    // One such example would be a code snippet, where we want to show that someone
+    // posted a code snippet but we might want to filter out the body.
+    /// The event rendered as a string.
+    pub message: String,
+}
+
+/// Trait allowing events to be rendered for Weechat.
+pub trait Render {
+    /// The event specific tags that should be attached to the rendered event.
+    const TAGS: &'static [&'static str];
+
+    /// Some events might need additional context to be rendered, for example if
+    /// we want to display the sender, we don't want to display the mxid,
+    /// instead we want to display the disambiguated display name.
```

Comment:
```suggestion
    /// Some events might need additional context to be rendered. For example,
    /// instead of displaying the MXID for the sender, we might want to display
    /// the disambiguated display name, which isn't available in the event.
```